tsdb/index/index.go

```diff
@@ -1414,34 +1408,41 @@ func (s symbolsIter) Err() error { return s.err }
 
 // ReadOffsetTable reads an offset table and at the given position calls f for each
 // found entry. If f returns an error it stops decoding and returns the received error.
-func ReadOffsetTable(bs ByteSlice, off uint64, f func([]string, uint64, int) error) error {
+func ReadOffsetTable[T any](bs ByteSlice, off uint64, read func(d *encoding.Decbuf) (T, error), f func(T, uint64, int) error) error {
 	d := encoding.NewDecbufAt(bs, int(off), castagnoliTable)
 	startLen := d.Len()
 	cnt := d.Be32()
 
 	for d.Err() == nil && d.Len() > 0 && cnt > 0 {
 		offsetPos := startLen - d.Len()
-		keyCount := d.Uvarint()
-		// The Postings offset table takes only 2 keys per entry (name and value of label),
-		// and the LabelIndices offset table takes only 1 key per entry (a label name).
-		// Hence setting the size to max of both, i.e. 2.
-		keys := make([]string, 0, 2)
 
-		for i := 0; i < keyCount; i++ {
-			keys = append(keys, d.UvarintStr())
+		item, err := read(&d)
+		if err != nil {
+			return err
 		}
 		o := d.Uvarint64()
 		if d.Err() != nil {
 			break
 		}
-		if err := f(keys, o, offsetPos); err != nil {
+		if err := f(item, o, offsetPos); err != nil {
 			return err
 		}
 		cnt--
 	}
 	return d.Err()
 }
 
+// PostingTableReader can be used as read function for ReadOffsetTable to read the posting offset table.
+func PostingTableReader(d *encoding.Decbuf) (labels.Label, error) {
```

Comment:
Should we call it `PostingsOffsetTableReader` (with the "s") for consistency?

Reply:
Done in https://github.com/prometheus/prometheus/pull/11535/commits/40cd01a10f83f216f750addeb8ee87fb045539ac